okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,245 @@
+import org.slf4j.LoggerFactory;
+
+public final class RemotingOkHttpCall extends ForwardingCall {
+
+    private static final Logger log = LoggerFactory.getLogger(RemotingOkHttpCall.class);
+
+    private static final ResponseHandler<RemoteException> remoteExceptionHandler =
+            RemoteExceptionResponseHandler.INSTANCE;
+    private static final ResponseHandler<IOException> catchAllExceptionHandler = IoExceptionResponseHandler.INSTANCE;
+    private static final ResponseHandler<QosException> qosHandler = QosExceptionResponseHandler.INSTANCE;
+
+    private final BackoffStrategy backoffStrategy;
+    private final UrlSelector urls;
+    private final RemotingOkHttpClient client;
+    private final ScheduledExecutorService schedulingExecutor;
+    private final ExecutorService executionExecutor;
+    private final Duration timeout;
+
+    private final int maxNumRelocations;
+
+    RemotingOkHttpCall(
+            Call delegate,
+            BackoffStrategy backoffStrategy,
+            UrlSelector urls, RemotingOkHttpClient client,
+            ScheduledExecutorService schedulingExecutor,
+            ExecutorService executionExecutor,
+            Duration timeout, int maxNumRelocations) {
+        super(delegate);
+        this.backoffStrategy = backoffStrategy;
+        this.urls = urls;
+        this.client = client;
+        this.schedulingExecutor = schedulingExecutor;
+        this.executionExecutor = executionExecutor;
+        this.timeout = timeout;
+        this.maxNumRelocations = maxNumRelocations;
+    }
+
+    private void scheduleExecution(Callback callback, Duration backoff) {
+        schedulingExecutor.schedule(
+                () -> executionExecutor.submit(() -> doClone().enqueue(callback)),
+                backoff.toMillis(),
+                TimeUnit.MILLISECONDS);
+    }
+
+    private void doEnqueueAndHandleErrorsAndRetries(Callback callback) {
+        super.enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException exception) {
+                if (!backoffStrategy.nextBackoff().isPresent()) {
+                    callback.onFailure(call, new IOException("Failed to complete the request due to an "
+                            + "IOException", exception));
+                } else {
+                    Optional<HttpUrl> redirectTo = urls.redirectToNext(request().url());
+                    if (!redirectTo.isPresent()) {
+                        callback.onFailure(call, new IOException("Failed to determine valid failover URL"
+                                + "for '" + request().url() + "' and base URLs " + urls.getBaseUrls()));
+                    } else {
+                        Request redirectedRequest = request().newBuilder()
+                                .url(redirectTo.get())
+                                .build();
+                        client.newCallWithMutableState(redirectedRequest, backoffStrategy, maxNumRelocations - 1)
```

Comment:
I think this backoffStrategy argument is the one that confuses me - the method says we're mutating stuff, but it appears that we're passing the same thing in again.  If it's internally mutable, we might as well leave it where it is - otherwise, I'd expect to be calling `.backoffMore()` or something

Reply:
BackoffStrategy.nextBackoff mutates the internal state.